### PR TITLE
feat(#2517): Add AI embeddings preprocessor

### PR DESCRIPTION
**Pull request by david-fritz-gravwell**
This PR addresses https://github.com/gravwell/issues/issues/2517

## This PR proposes...

- Adding an AI embeddings preprocessor that attaches an intrinsic EV with the vector. 
- This is intended to be used with an upcoming semantic search module.

### CODE REVIEW

File: ingest/processors/vector.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	VectorProcessor string = `vector`
+
+	defaultVectorTimeout       = 30 // seconds
+	defaultVectorRetryAttempts = 3
+	vectorRetryBaseDelay       = 250 * time.Millisecond
+	vectorRetryMaxDelay        = 5 * time.Second
+)
+
+var (
+	ErrMissingModel    = errors.New("Missing Model in vector config")
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go]: In my editor, staticheck gives me the error:

`1. ST1005: error strings should not be capitalized [default]`

I realize this is everywhere in the codebase already, but let's try no to add to it.

File: ingest/processors/vector.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+		return ErrMissingEndpoint
+	}
+	if u, err := url.Parse(c.Endpoint); err != nil || (u.Scheme != "http" && u.Scheme != "https") || u.Host == "" {
+		return ErrInvalidEndpoint
+	}
+	return nil
+}
+
+type VectorProc struct {
+	nocloser
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go:81]: `nop_closer` is a bit clearer IMO, I had to look up what this was to understand, while `nop_closer` is pretty obvious.

File: ingest/processors/vector.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+	// Collect the non-nil entries and their payloads so the whole batch can be
+	// embedded in a single request.
+	idx := make([]int, 0, len(ents))
+	inputs := make([]string, 0, len(ents))
+	for i, ent := range ents {
+		if ent == nil {
+			continue
+		}
+		idx = append(idx, i)
+		inputs = append(inputs, string(ent.Data))
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go:134]: What about binary data? I realize it would be stupid of someone to try to get vec embeddings of non printable string data, but It could happen... Is there anything we could do? What does the model do?

File: ingest/processors/vector.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+	}
+	rset = ents[:0]
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go:137]: Could you put a comment explaining this line and the intention behind it? We had a whole discussion on the backend channel for this, which is a good indication it should at least be explained for posterity

File: ingest/processors/vector.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+		return nil, false, ErrEmbeddingCount
+	}
+
+	// The API may return data out of order; place each by its index.
+	embeddings := make([][]float64, want)
+	for _, d := range embResp.Data {
+		if d.Index < 0 || d.Index >= want {
+			return nil, false, ErrEmbeddingCount
+		}
+		if len(d.Embedding) == 0 {
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go:265]: This is probably the answer to: https://github.com/gravwell/gravwell/pull/2493/changes#r3498987540

However... do we want to stop all embeddings if one of the entries was unprocessable? Shouldn't we try to embed what we can?

File: ingest/processors/vector_test.go [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(rset) != 1 {
+		t.Fatalf("expected success after retries, got %d results", len(rset))
+	}
+	if got := atomic.LoadInt32(&calls); got != 3 {
+		t.Errorf("expected 3 attempts (2 rate-limited + 1 success), got %d", got)
+	}
+}
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector_test.go:425]: I'm missing some test cases here, like the error paths of `doEmbeddingRequest` (empty embeddings / wrong count)

**Review by daniel-moreno-levy-gravwell — CHANGES_REQUESTED** [review]

File: ingest/processors/vector.go (hunk 2 above)
Comment (daniel-moreno-levy-gravwell) [reply ingest/processors/vector.go:81]: Additionally, it would be nice to leave a comment here as to why we these methods are necessary... It took me some digging to see that `Processor` interface requires it.

Comment (david-fritz-gravwell) [reply ingest/processors/vector.go:81]: The `nocloser` type comes from the processors definition. Won't fix/change here.

File: ingest/processors/vector.go (hunk 3 above)
Comment (david-fritz-gravwell) [reply ingest/processors/vector.go:134]: Strings only -- and this should be documented for sure.

File: ingest/processors/vector.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+	embeddings, eerr := vp.getEmbeddings(inputs)
+	if eerr != nil {
+		// In passthrough mode we keep going quietly and emit the originals
+		// untouched. Otherwise surface the error so the ingester logs it; the
+		// batch is dropped.
+		if vp.Passthrough_On_Error {
+			return ents, nil
+		}
+		return nil, eerr
+	}
```
[the remark this answers is not in the document]
Comment (david-fritz-gravwell) [reply ingest/processors/vector.go:151]: We have to filter nils on success because that's how embeddings work, but nil entries are otherwise valid.

File: ingest/processors/vector_test.go (hunk 6 above)
Comment (david-fritz-gravwell) [reply ingest/processors/vector_test.go:425]: Done.

File: ingest/processors/vector.go (hunk 4 above)
Comment (david-fritz-gravwell) [reply ingest/processors/vector.go:137]: Done

File: ingest/processors/vector.go (hunk 1 above)
Comment (david-fritz-gravwell) [reply ingest/processors/vector.go]: Done.

File: ingest/processors/vector.go (hunk 5 above)
Comment (david-fritz-gravwell) [reply ingest/processors/vector.go:265]: I'm not sure what the right thing to do here is. I don't know if it's possible for a model to respond with "some but not all" embeddings, so for now at least I think this is the behavior I want...

File: ingest/processors/vector.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+	vectorRetryMaxDelay        = 5 * time.Second
+	maxEmbeddingResponseSize   = 1 << 20 // 1 MB — guard against unbounded response bodies
+)
+
+var (
+	ErrMissingModel    = errors.New("missing model in vector config")
+	ErrMissingEndpoint = errors.New("missing endpoint in vector config")
+	ErrInvalidEndpoint = errors.New("invalid endpoint in vector config")
+	ErrEmptyEmbedding  = errors.New("empty embedding response")
+	ErrEmbeddingCount  = errors.New("Embedding response count does not match request")
```
Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go]: Initial uppercase error..

Comment (daniel-moreno-levy-gravwell) [inline ingest/processors/vector.go]:
```suggestion
	ErrEmbeddingCount  = errors.New("embedding response count does not match request")
```

**Review by daniel-moreno-levy-gravwell — CHANGES_REQUESTED** [review]
Nitpik, everything else was either addressed or given a good reason not to. After the small change request, it LGTM.

**Review by daniel-moreno-levy-gravwell — APPROVED** [review]
